source/robot_model/src/Model.cpp

```diff
@@ -5,277 +5,381 @@
-  jacobian.bottomRows(6) = this->compute_jacobian(joint_state, state.get_name()).data();
-  // compute the Jacobian
-  Eigen::MatrixXd hessian_matrix = jacobian.transpose() * jacobian;
-
-  // set the hessian sparse matrix
-  std::vector<Eigen::Triplet<double>> coefficients;
-  for (unsigned int i = 0; i < nb_joints; ++i) {
-    for (unsigned int j = 0; j < nb_joints; ++j) {
-      coefficients.emplace_back(Eigen::Triplet<double>(i, j, hessian_matrix(i, j)));
+
+  StateRepresentation::JointPositions Model::inverse_geometry(StateRepresentation::JointState actual_joint_state,
+                                                              const StateRepresentation::CartesianState &desired_cartesian_state)
+  {
+    // This method uses the angular velocity evaluated by the 'Infeasibility-free inverse kinematics method' which takes into account the
+    // joint limits. The angular velocity is then used for indicating the direction for an iterative algorithm.
+    // To converge at the desired end-effector position a similar approach to the Newton–Raphson method is used, the only difference is that
+    // here the pseudo-inverse of the jacobian is not used and instead is used the angualr velocity evaluated by the
+    // 'Infeasibility-free inverse kinematics method'.
+
+    const int JOINT_ID = this->robot_model_.njoints - 2;
+
+    // TODO if the actual_joint_state is not provided initialize the position of the robot in a radom conformation
+    // TODO take as input the parameters eps and IT_MAX
+    // TODO return 'success'
+
+    Eigen::VectorXd q = actual_joint_state.get_positions(); //pinocchio::neutral(model);
+    const double eps = 1e-4;
+    const int IT_MAX = 1000;
+
+    std::vector<unsigned int> frame_ids{(unsigned int)JOINT_ID};
+
+    StateRepresentation::CartesianPose actual_cartesian_state;
+    double error;
+
+    bool success = false;
+    Eigen::VectorXd v(this->robot_model_.nv);
+    for (int i = 0;; i++)
+    {
+
+      actual_cartesian_state = this->forward_geometry(actual_joint_state, frame_ids).front();
+      error = desired_cartesian_state.dist(actual_cartesian_state, StateRepresentation::CartesianStateVariable::POSE);
+
+      if (error < eps)
+      {
+        success = true;
+        break;
+      }
+      if (i >= IT_MAX)
+      {
+        success = false;
+        break;
+      }
+      v = this->inverse_kinematic(actual_joint_state, desired_cartesian_state).get_velocities();
+      q = actual_joint_state.get_positions();
+      actual_joint_state.set_velocities(v);
+      StateRepresentation::CartesianTwist x_dot = this->forward_kinematic(actual_joint_state);
+
+      StateRepresentation::CartesianTwist zerovec;
+      zerovec.Zero("zero");
+      double x_dot_norm = x_dot.dist(zerovec, StateRepresentation::CartesianStateVariable::POSE);
+
+      // dt=||x_d - x||/||x'||
+      double dt = error / x_dot_norm;
+
+      q = pinocchio::integrate(this->robot_model_, q, v * dt);
```

Comment:
You should use the operations from `JointState` to do this. Keep everything in `JointVelocities` and `JointPositions`

Reply:
oh ok not quite so as you are passing `actual_joint_state` by value and not by reference. Please pass `actual_joint_state` as a const reference and instead make a copy of it if needed

Reply:
I used the `pinocchio::integrate` because I guess it takes into account the joint limits.

Reply:
Oh ok are you sure? Then yes it is a neat feature but should be made explicit

Reply:
I got inspired by the IK (click) showed in the pinocchio documentation https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/md_doc_b-examples_i-inverse-kinematics.html
But I think I am wrong because its input should be a velocity and I put v*dt. I will modify that part and I will put something that checks the joint limits.

Reply:
In this version I am still using `integrate` but if you prefer I can change it. For now I have tested this version and it seams working